course/assets/js/video-download-button.js

```diff
@@ -0,0 +1,86 @@
+//@ts-nocheck
+import videojs from "video.js"
+
+function _createDownloadMenuItems() {
+  const MenuItems = videojs.getComponent("MenuItem")
+  const DownloadMenuItems = videojs.extend(MenuItems, {
+    constructor: function(player, options) {
+      options.selectable = false
+      MenuItems.call(this, player, options)
+    },
+    handleClick: function() {
+      this.options_.callable()
+    }
+  })
+  MenuItems.registerComponent("DownloadMenuItems", DownloadMenuItems)
+}
+
+function _createDownloadbutton() {
+  const MenuButton = videojs.getComponent("MenuButton")
+  const DownloadMenuButton = videojs.extend(MenuButton, {
+    constructor: function(player, options) {
+      this.label = document.createElement("span")
+      this.values = options.values
+      options.label = "download options"
+
+      MenuButton.call(this, player, options)
+      this.$("button").classList.add("vjs-download-button")
+      this.el().setAttribute("aria-label", "download button")
+      this.controlText("download button")
+    },
+    createItems: function() {
+      const menuItems = []
+      const DownloadMenuItems = videojs.getComponent("DownloadMenuItems")
+      this.values.map(item => {
+        menuItems.push(
+          new DownloadMenuItems(this.player_, {
+            label:    item[0],
+            callable: item[1]
+          })
+        )
+      })
+      return menuItems
+    }
+  })
+  MenuButton.registerComponent("DownloadMenuButton", DownloadMenuButton)
+}
+
+export function initDownloadButton() {
+  _createDownloadMenuItems()
+  _createDownloadbutton()
+
+  const DownloadMenuButton = videojs.getComponent("DownloadMenuButton")
+
+  if (document.querySelector(".video-container")) {
+    const videoPlayers = document.querySelectorAll(".vjs-ocw")
+    for (const videoPlayer of Array.from(videoPlayers)) {
```

Comment:
I'm thinking of the reason for which we have a loop here, most probably for a use-case in which we might have multiple `video players`, so keeping this use-case in mind: shouldn't we have different and multiple `window.videoDownloadLink` and `window.transcriptDownloadLink`, (each video should have its own downloadable link)?
Or this loop is for multiple video players of the same video but different tracks/speeds etc? 

Currently, if we have multiple video players, all of them will have the same downloadable links.

Maybe one approach can be to store an array of objects consisting of `youtube keys`/`video player ID` and downloadable links:
something like

```
window.videoTranscriptDownloadLinks = [
  { 
    videoPlayerID1: {
      videoDownloadLink: 'url',
      transcriptDownloadLink: 'url',
    }  
  }, 
  { 
    videoPlayerID2: {
      videoDownloadLink: 'url',
      transcriptDownloadLink: 'url',
    }  
  }, 
]
```
we can also have 2 arrays: 1 for video download links and 1 for transcript download links.

Another approach, which I like more, can be to store the download links in `<video>` tag as an HTML data attribute:
Something like: 
```
<video
  id="video-player-{{ .youtubeKey }}-{{ $random }}"
  class="video-js vjs-default-skin vjs-big-play-centered vjs-ocw"
  data-videoDownloadLink: "url"
  data-transcriptDownloadLink: "url"
  .....
>
```
and use it in the JS file when getting the videoplayer.

but since I'm not very much sure about this multiple video-players use-case, so its better to get someone's else thoughts before implementing a solution for it, though I'm sure we can have multiple embedded videos.

Maybe, @gumaerc or @abeglova might have better insight on this.

Reply:
Yes, we sometimes have multiple video players in one page for embedded videos, Peter linked to one in slack. Those don't have transcript or video download links. The resource page for a video always only has one video player